src/main/java/baritone/pathing/movement/MovementHelper.java

```diff
@@ -88,61 +88,105 @@ static boolean canWalkThrough(BlockStateInterface bsi, int x, int y, int z) {
-        if (block == Blocks.AIR) { // early return for most common case
-            return true;
+
+        if (block == Blocks.AIR) {
+            return Optional.of(true);
         }
+
         if (block == Blocks.FIRE || block == Blocks.TRIPWIRE || block == Blocks.WEB || block == Blocks.END_PORTAL || block == Blocks.COCOA || block instanceof BlockSkull || block instanceof BlockTrapDoor || block == Blocks.END_ROD) {
-            return false;
+            return Optional.of(false);
         }
+
         if (Baritone.settings().blocksToAvoid.value.contains(block)) {
-            return false;
+            return Optional.of(false);
         }
+
         if (block instanceof BlockDoor || block instanceof BlockFenceGate) {
             // Because there's no nice method in vanilla to check if a door is openable or not, we just have to assume
             // that anything that isn't an iron door isn't openable, ignoring that some doors introduced in mods can't
             // be opened by just interacting.
-            return block != Blocks.IRON_DOOR;
+            return Optional.of(block != Blocks.IRON_DOOR);
         }
+
         if (block == Blocks.CARPET) {
-            return canWalkOn(bsi, x, y - 1, z);
+            return Optional.empty();
         }
+
         if (block instanceof BlockSnow) {
-            // we've already checked doors and fence gates
-            // so the only remaining dynamic isPassables are snow and trapdoor
-            // if they're cached as a top block, we don't know their metadata
-            // default to true (mostly because it would otherwise make long distance pathing through snowy biomes impossible)
-            if (!bsi.worldContainsLoadedChunk(x, z)) {
-                return true;
-            }
-            // the check in BlockSnow.isPassable is layers < 5
-            // while actually, we want < 3 because 3 or greater makes it impassable in a 2 high ceiling
             if (state.getValue(BlockSnow.LAYERS) >= 3) {
-                return false;
+                return Optional.of(false);
             }
-            // ok, it's low enough we could walk through it, but is it supported?
+
+            return Optional.empty();
+        }
+
+        if (block instanceof BlockLiquid) {
+            if (state.getValue(BlockLiquid.LEVEL) != 0) {
+                return Optional.of(false);
+            } else {
+                return Optional.empty();
+            }
+        }
+
+        if (block instanceof BlockCauldron) {
+            return Optional.of(false);
+        }
+
+        try { // A dodgy catch-all at the end, for most blocks with default behaviour this will work, however where blocks are special this will error out, and we can handle it when we have this information
+            return Optional.of(block.isPassable(null, null));
+        } catch (NullPointerException exception) {
+            System.out.println("Error");
+            return Optional.empty();
```

Comment:
It took me a second to realize why this is correct - let's make this comment a bit more explanatory. If I'm understanding right, the idea is that `block.isPassable(bsi.access, bsi.isPassableBlockPos.setPos(x, y, z));` can only be done in special case, not in precomputed, so if we default to `Optional.empty()` then we're good because the special case will call it correctly?

Reply:
Also just out of kindness to other mods that will lose their minds when passing null data into `block.isPassable`, perhaps this should be a general `catch Throwable`?

Reply:
Ok, I can switch it to a throwable. The idea is that for most common blocks it won't crash when doing this, most don't override the default `return !this.material.blocksMovement();` however some do, and in those cases they probably use context from the world, which means we have to calculate them as a special case.